File: src/cc/Common/MetricsProcess.cc
/*
 * Copyright (C) 2007-2014 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3
 * of the License.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for MetricsProcess.
/// This file contains type declarations for MetricsProcess, a simple class for
/// aggregating metrics and sending them to the Ganglia gmond process running on
/// localhost.

#include <Common/Compat.h>

#include "MetricsProcess.h"

#include <Common/Logger.h>
#include <Common/StatsSystem.h>
#include <Common/Time.h>

using namespace Hypertable;
using namespace std;

MetricsProcess::MetricsProcess() {
  StatsSystem system_stats(StatsSystem::PROCINFO|StatsSystem::PROC);
  system_stats.refresh();
  m_last_timestamp = get_ts64();
  m_last_sys = system_stats.proc_stat.cpu_sys;
  m_last_user = system_stats.proc_stat.cpu_user;
}

void MetricsProcess::collect(int64_t now, MetricsCollector *collector) {
  StatsSystem system_stats(StatsSystem::PROCINFO|StatsSystem::PROC);

  system_stats.refresh();

  int64_t elapsed_millis = (now - m_last_timestamp) / 1000000LL;
  int64_t diff_sys = (system_stats.proc_stat.cpu_sys - m_last_sys) / System::cpu_info().total_cores;
  int64_t diff_user = (system_stats.proc_stat.cpu_user - m_last_user) / System::cpu_info().total_cores;
  int32_t pct = 0;

  // CPU sys
  if (elapsed_millis)
    pct = (diff_sys * 100) / elapsed_millis;
  collector->update("cpu.sys", pct);

  // CPU user
  if (elapsed_millis)
    pct = (diff_user * 100) / elapsed_millis;
  collector->update("cpu.user", pct);

  // Virtual memory
  collector->update("memory.virtual",
                    (float)system_stats.proc_stat.vm_size / 1024.0);

  // Resident memory
  collector->update("memory.resident",
                    (float)system_stats.proc_stat.vm_resident / 1024.0);

  // Heap size
  collector->update("memory.heap",
                    (float)system_stats.proc_stat.heap_size / 1000000000.0);

  // Heap slack bytes
  collector->update("memory.heapSlack",
                    (float)system_stats.proc_stat.heap_slack / 1000000000.0);

  // Hypertable version
  collector->update("version", version_string());

  m_last_timestamp = now;
  m_last_sys = system_stats.proc_stat.cpu_sys;
  m_last_user = system_stats.proc_stat.cpu_user;

}

File: java/hypertable-common/src/main/java/org/hypertable/Common/MetricsProcess.java
/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Declaration for MetricsProcess.
 * This file contains the class definition of MetricsProcess, a class for
 * computing process metrics.
 */

package org.hypertable.Common;

import java.lang.Long;
import java.lang.Runtime;
import java.lang.System;
import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.ManagementFactory;
import java.util.HashMap;

import org.hyperic.sigar.jmx.SigarProcess;
import org.hyperic.sigar.Mem;
import org.hyperic.sigar.Sigar;
import org.hyperic.sigar.SigarException;

/** Collects process metrics. */
public class MetricsProcess {

  /** Constructor.
   * Initializes mLastSys and mLastUser from data gathered from SIGAR.
   * Initializes mLastTimestamp to current time in milliseconds.
   */
  public MetricsProcess() {
    SigarProcess process = new SigarProcess(mSigar);
    mLastSys = process.getTimeSys();
    mLastUser = process.getTimeUser();
    mLastTimestamp = System.currentTimeMillis();
  }

  /** Collects process metrics.
   * Computes process metrics and publishes them via <code>collector</code>.
   * The following JVM metrics and process metrics collected from SIGAR are
   * computed and published:
   * <ul>
   * <li> <code>jvm.gc</code> - GC count</li>
   * <li> <code>jvm.gcTime</code> - GC time (ms)</li>
   * <li> <code>jvm.heapSize</code> - JVM heap size (GB)</li>
   * <li> <code>cpu.sys</code> - CPU system time (%)</li>
   * <li> <code>cpu.user</code> - CPU user time (%)</li>
   * <li> <code>memory.virtual</code> - Virtual memory (GB)</li>
   * <li> <code>memory.resident</code> - Resident memory (GB)</li>
   * </ul>
   * @param now Current time in milliseconds
   * @param collector Metrics collector
   */
  public void collect(long now, MetricsCollector collector) {
    HashMap<GarbageCollectorMXBean, Long> collectionCounts = new HashMap<GarbageCollectorMXBean, Long>();
    HashMap<GarbageCollectorMXBean, Long> collectionTimes = new HashMap<GarbageCollectorMXBean, Long>();
    long totalGarbageCollections = 0;
    long garbageCollectionTime = 0;

    for(GarbageCollectorMXBean gc :
          ManagementFactory.getGarbageCollectorMXBeans()) {

      long count = gc.getCollectionCount();
      collectionCounts.put(gc, count);
      if (mCollectionCounts.containsKey(gc)) {
        count -= mCollectionCounts.get(gc).longValue();
      }
      if (count >= 0) {
        totalGarbageCollections += count;
      }

      long time = gc.getCollectionTime();
      collectionTimes.put(gc, time);
      if (mCollectionTimes.containsKey(gc)) {
        time -= mCollectionTimes.get(gc).longValue();
      }
      if(time >= 0) {
        garbageCollectionTime += time;
      }
    }

    /* JVM GCs */
    collector.update("jvm.gc", (int)totalGarbageCollections);

    /* JVM GC time */
    collector.update("jvm.gcTime", (int)garbageCollectionTime);

    /* JVM heap size */
    double heapGb = (double)Runtime.getRuntime().totalMemory() / 1000000000.0;
    collector.update("jvm.heapSize", heapGb);

    SigarProcess process = new SigarProcess(mSigar);

    double gb;
    long cpuTimeSys = process.getTimeSys();
    long cpuTimeUser = process.getTimeUser();
    long elapsed_millis = now - mLastTimestamp;
    long diff_sys = (cpuTimeSys - mLastSys) / org.hypertable.Common.System.processorCount;
    long diff_user = (cpuTimeUser - mLastUser) / org.hypertable.Common.System.processorCount;

    int pct = 0;

    /* CPU sys */
    if (elapsed_millis > 0)
      pct = (int)((diff_sys * 100) / elapsed_millis);
    collector.update("cpu.sys", pct);

    /* CPU user */
    if (elapsed_millis > 0)
      pct = (int)((diff_user * 100) / elapsed_millis);
    collector.update("cpu.user", pct);

    /* Virtual memory */
    gb = (double)process.getMemVsize() / (double)(1024*1024*1024);
    collector.update("memory.virtual", gb);

    /* Resident memory */
    gb = (double)process.getMemResident() / (double)(1024*1024*1024);
    collector.update("memory.resident", gb);

    mCollectionCounts = collectionCounts;
    mCollectionTimes = collectionTimes;
    mLastTimestamp = now;
    mLastSys = cpuTimeSys;
    mLastUser = cpuTimeUser;
  }

  /** Map of last GC collection counts */
  HashMap<GarbageCollectorMXBean, Long> mCollectionCounts = new HashMap<GarbageCollectorMXBean, Long>();

  /** Map of last GC collection times */
  HashMap<GarbageCollectorMXBean, Long> mCollectionTimes = new HashMap<GarbageCollectorMXBean, Long>();

  /** SIGAR */
  private static Sigar mSigar = new Sigar();

  /** Last collection timestamp in milliseconds */
  long mLastTimestamp = 0;

  /** Last recorded CPU system time */
  long mLastSys = 0;

  /** Last recorded CPU user time */
  long mLastUser = 0;

}
